### PR TITLE
tabular kaggle tutorial

### PR DESCRIPTION
**Pull request by jwmueller**
*Description of changes:*

Notebook tutorial for running autogluon on Kaggle competitions with tabular data. This tutorial is slow since Kaggle dataset is very large, so it is not executed with {.python .input} blocks.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-166-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-166/1/index.html

**Comment by szha** [comment]
Job PR-166-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-166/2/index.html

**Comment by szha** [comment]
Job PR-166-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-166/3/index.html

**Comment by szha** [comment]
Job PR-166-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-166/4/index.html

**Comment by szha** [comment]
Job PR-166-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-166/6/index.html

**Comment by szha** [comment]
Job PR-166-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-166/7/index.html

File: docs/tutorials/tabular_prediction/tabular-kaggle.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+`kaggle competitions submit -c ieee-fraud-detection -f sample_submission.csv -m "my first submission"`
+
+You can now play with different `fit()` arguments and feature-engineering techniques to try and maximize the rank of your submissions in the Kaggle Leaderboard!
+
+
+**Tips to maximize predictive performance:**
+
+   - Be sure to specify the appropriate evaluation metric if one is specified on the competition website! If you are unsure which metric is best, then simply do not specify this argument when invoking `fit()`; AutoGluon should still produce high-quality models by automatically inferring which metric to use.
+
+   - If the training examples are time-based and the competition test examples come from future data, we recommend you reserve the most recently-collected training examples as a separate validation dataset passed to `fit()`. Otherwise, you do not need to specify a validation set yourself and AutoGluon will automatically partition the competition training data into its own training/validation sets.
```
Comment (Innixma) [inline docs/tutorials/tabular_prediction/tabular-kaggle.md]: Mention that they should avoid bagging and stacking if this is the case, since we will use the val data in bagging currently.

File: tests/unittests/test_tabular.py [hunk 2]
```diff
@@ -28,75 +28,93 @@
...
+    # If True, run a faster benchmark (subsample training sets, less epochs, etc),
+    # otherwise we run full benchmark with default AutoGluon settings.
+    # performance_value warnings are disabled when fast_benchmark = True.
+
+    #### If fast_benchmark = True, can control model training time here. Only used if fast_benchmark=True ####
+    if fast_benchmark:
+        subsample_size = 100
+        nn_options = {'num_epochs': 3} 
+        gbm_options = {'num_boost_round': 30}
+        hyperparameters = {'GBM': gbm_options, 'NN': nn_options}
```
Comment (Innixma) [inline tests/unittests/test_tabular.py]: Maybe we should start using the default hyperparameters to ensure all default models are working properly? This test wouldn't catch a crash in KNN for example.

File: tests/unittests/test_tabular_bagstack.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    if hyperparameters is not None:
+        fit_args['hyperparameters'] = hyperparameters
+    if time_limits is not None:
+        fit_args['time_limits'] = time_limits
+    ###################################################################
+    run_tabular_benchmarks(fast_benchmark=fast_benchmark, subsample_size=subsample_size, perf_threshold=perf_threshold,
+                           seed_val = seed_val, fit_args=fit_args)
+
+
+def test_tabular_stack2():
```
Comment (Innixma) [inline tests/unittests/test_tabular_bagstack.py]: This is thorough but perhaps we could just check if the stack_ensemble_levels is set to 0 because of num_bagging_folds=0 instead of actually running it. We could do this by splitting .fit internal logic into two stages: Handling params and the call to learner.fit(), we could avoid running the second stage and just check if stack_ensemble_levels was set to 0.

File: tests/unittests/test_tabular_bagstack.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    if hyperparameters is not None:
+        fit_args['hyperparameters'] = hyperparameters
+    if time_limits is not None:
+        fit_args['time_limits'] = time_limits
+    ###################################################################
+    run_tabular_benchmarks(fast_benchmark=fast_benchmark, subsample_size=subsample_size, perf_threshold=perf_threshold,
+                           seed_val=seed_val, fit_args=fit_args)
+
+
+def test_tabular_stack1():
```
Comment (Innixma) [inline tests/unittests/test_tabular_bagstack.py]: ditto to below comment

File: tests/unittests/test_tabular_bagstack.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    if hyperparameters is not None:
+        fit_args['hyperparameters'] = hyperparameters
+    if time_limits is not None:
+        fit_args['time_limits'] = time_limits
+    ###################################################################
+    run_tabular_benchmarks(fast_benchmark=fast_benchmark, subsample_size=subsample_size, perf_threshold=perf_threshold,
+                           seed_val=seed_val, fit_args=fit_args)
+
+
+if __name__ == '__main__':
```
Comment (Innixma) [inline tests/unittests/test_tabular_bagstack.py]: Seems like a lot of duplicate code here. I think it might be nice to reduce the duplication by having default parameter values and then only having to specify the non-standard parameters to the test func (stack_ensemble_levels and num_bagging_folds). This would avoid subtle bugs if we have to refactor this later.

This could be done by having another function run_tabular_benchmarks_template(), which would call run_tabular_benchmarks with all the code above, but having you able to specify overrides.

File: docs/tutorials/tabular_prediction/tabular-kaggle.md (hunk 1 above)
Comment (jwmueller) [reply docs/tutorials/tabular_prediction/tabular-kaggle.md]: Done in new commit.  I guess we also needed to mention this fact in the fit() docs, so have done so as well.

File: tests/unittests/test_tabular.py (hunk 2 above)
Comment (jwmueller) [reply tests/unittests/test_tabular.py]: will add one unit-test that runs with default hyperparams. The problem is default may be slow, depending on models/dataset

File: tests/unittests/test_tabular_bagstack.py (hunk 3 above)
Comment (jwmueller) [reply tests/unittests/test_tabular_bagstack.py]: you don't think we'd ever support stacking without bagging in the future? If so I can just remove these tests {test_tabular_stack1() and test_tabular_stack2()} from being run at the moment.

File: tests/unittests/test_tabular_bagstack.py (hunk 5 above)
Comment (jwmueller) [reply tests/unittests/test_tabular_bagstack.py]: I actually think we eventually want to change the defaults in the different tests, so we have more comprehensive test coverage of various scenarios. This is why the parameters are defined in every test, they just happen to be the same right now.  I'm not sure our our set of default base models is fully finalized, but once it is, I think these parameters should aim cover different subsets of base models in different unit tests. Because I'm confident GBM and NN will be in the official final set of base models, this is why these are currently the only ones tried for now.  I've gone and changed the seed for the different tests in new commit.

**Comment by szha** [comment]
Job PR-166-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-166/9/index.html
